### ISSUE TITLE
Add a lazy SliverTree

### ISSUE THREAD

**Issue by naamapps** [flutter/flutter#114299]
📜  [Design Document](https://docs.google.com/document/d/1-aFI7VjkF9yMkWpP94J8T_JREDS-M3bOak26PVehUYg/edit?usp=sharing)

---

Related:
- https://github.com/flutter/flutter/issues/31606

## Use case
Currently when trying to implement tree UIs (for example: comments section with nested comments - see Instagram's comments for reference) there is no way to do it but to flatten the list and use one SliverList to show the data in the UI.
## Proposal
Add a way to add a SliverList inside of a SliverList.
This will allow the CustomListView to scroll with all the data while also perform nicely with all the Slivers benefits.

In my use case, this is critical in SliverAnimatedList.

If you have any other solutions I will be happy to discuss.

## Solutions I tried
1. Using a ListView.Builder inside the SliverList builder function with shrinkWrap: true.  - **Poor Performance**
2. As pointed above - Normalizing the nested list with the list.expand function and using this list as the data source. - **Poor Developer Experience** - because everytime new data arrives we need to manage the flattened list which is not ideal.

## Code example
```dart
CustomScrollView(
    slivers: [
      SliverList(
        delegate: SliverChildBuilderDelegate(
          (context, index) {
            final data = widget.comments[index];
            return SliverList(
              delegate: SliverChildBuilderDelegate(
                (context, index) {
                  if (index == 0) {
                    return CommentWidget(data);
                  } else {
                    return CommentWidget(
                        data.comments[index - 1],
                        isNested: true);
                  }
                },
                childCount: data.comments.length + 1,
              ),
            );
          },
          childCount: widget.comments.length,
        ),
      ),
    ]
)
```

**Comment by jonahwilliams** [comment]
Not quite SliverList, but this would be called something like SliverGroup. I don't believe we've writen it yet. FYI @Piinks if we have a bug for this already

**Comment by naamapps** [comment]
Thanks for the quick response @jonahwilliams,
Just to clarify, we need a widget that gets a builder function in order to get the most performence out of this feature.
Adding a widget that gets a children array is not quite helpful here since we can have plenty of nested items.

**Comment by naamapps** [comment]
Thanks @Piinks for the detailed information.

I think that SliverList supporting both box and slivers as children actually simplifies the API, don't you think?
It is more friendly for new flutter devs and also more intuitive in my opinion.. 

I prefer building my own solutions than using a widget with limited customizations.
I have set in my project an infra for scrolling (with plenty of customization - animations, states and more) and it would really complicate the code to add a new widget..

That is why I suggested to nest a SliverList inside a SliverList, just because it is logical and also easy to use and understand as a developer.

Nevertheless, I look forward to see your solutions!
Hope to hear from you soon,
Thanks
